doc/mpls/MPLS_hld.md

```diff
@@ -0,0 +1,510 @@
+
+For MPLS, RouteOrch was modified to monitor updates to the new APPL_DB LABEL_ROUTE_TABLE.  RouteOrch translates all updates to LABEL_ROUTE_TABLE to equivalent SAI requests for SAI MPLS API.
+NextHop processing for updates from both the new LABEL_ROUTE_TABLE and the existing ROUTE_TABLE has been extended to detect possible MPLS LabelStack content and propagate this additional information to NeighOrch for SAI handling.
+###### Functions
+The following are new functions for RouteOrch:
+```
+  /* Consumer handler for all events in APPL_DB LABEL_ROUTE_TABLE */
+  void RouteOrch::doLabelTask(Consumer& consumer);
+  /* Handler to process new MPLS route from LABEL_ROUTE_TABLE */
+  bool addLabelRoute(LabelRouteBulkContext& ctx, const NextHopGroupKey&);
+  /* Handler to process MPLS route removal from LABEL_ROUTE_TABLE */
+  bool removeLabelRoute(LabelRouteBulkContext& ctx);
+```
+
+##### NeighOrch
+NeighOrch is an existing component of the OrchAgent daemon in the SWSS container.  NeighOrch monitors operations on Neighbor related tables in APPL_DB.  Additionally NeighOrch coordinates Next Hop operations with RouteOrch and converts operations into SAI commands to manage Next Hop objects.
+
+For MPLS, NeighOrch has been extended to send create/remove SAI requests for MPLS NextHop objects (ie, NextHop objects that include non-empty LabelStack information) when associated Neighbor objects are created/removed. This MPLS NextHop behavior parallels the existing IPv4/IPv6 NextHop behavior in NeighOrch.
+###### Functions
+Existing functions from NeighOrch are updated to include NextHopKey parameter instead of IpAddress and visibility was raised to public for RouteOrch accessbility.
+```
+    bool hasNextHop(const NextHopKey&);
+    bool addNextHop(const NextHopKey&);
+    bool removeNextHop(const NextHopKey&);
+```
+
+##### CrmOrch
+CrmOrch is an existing component of the OrchAgent daemon in the SWSS container.  CrmOrch monitors resource usage in the SONiC system and triggers alarms when configurable thresholds are reached.
+
+For MPLS, CrmOrch has been extended to monitor the number of MPLS in-segment entries against the platform-specific number of entries available.  To facilitate this, a new CRM resource type CRM_MPLS_INSEG has been added to CrmOrch and mapped to the existing SAI object type SAI_OBJECT_TYPE_INSEG_ENTRY for querying via sai_object_type_get_availability().  This MPLS behavior parallels the existing IPv4/IPv6 behavior with CRM_IPV4_ROUTE/SAI_SWITCH_ATTR_AVAILABLE_IPV4_ROUTE_ENTRY and CRM_IPV6_ROUTE/SAI_SWITCH_ATTR_AVAILABLE_IPV6_ROUTE_ENTRY.
+###### Functions
+No new functions were required for CrmOrch MPLS support.
+
+##### Label/LabelStack
+Label and LabelStack are new type utilities of the OrchAgent daemon in the SWSS container.
+These types are introduced to represent the MPLS label or label stack when found in an in-segment entry or next hop.
```

Comment:
Is this being handled in new MPLS orch agent module?

Reply:
Can we add some details about how do we determine pop vs swap operation for an LSP in orchagent?

Reply:
the major change is in routeorch it's along with other types of routes(ipv4/v6 etc). The operation(pop/swap/push) is based on prefix and NH label stacks.